src/dask_nested/core.py

```diff
@@ -0,0 +1,385 @@
+from __future__ import annotations
+
+import os
+
+import dask.dataframe as dd
+import dask_expr as dx
+import nested_pandas as npd
+from dask_expr._collection import new_collection
+from nested_pandas.series.dtype import NestedDtype
+from nested_pandas.series.packer import pack_flat
+from pandas._libs import lib
+from pandas._typing import AnyAll, Axis, IndexLabel
+from pandas.api.extensions import no_default
+
+# need this for the base _Frame class
+# mypy: disable-error-code="misc"
+
+
+class _Frame(dx.FrameBase):  # type: ignore
+    """Base class for extensions of Dask Dataframes."""
+
+    _partition_type = npd.NestedFrame
+
+    def __init__(self, expr):
+        super().__init__(expr)
+
+    @property
+    def _args(self):
+        # Ensure our Dask extension can correctly be used by pickle.
+        # See https://github.com/geopandas/dask-geopandas/issues/237
+        return super()._args
+
+    def optimize(self, fuse: bool = True):
+        result = new_collection(self.expr.optimize(fuse=fuse))
+        return result
+
+    def __dask_postpersist__(self):
+        func, args = super().__dask_postpersist__()
+
+        return self._rebuild, (func, args)
+
+    def _rebuild(self, graph, func, args):  # type: ignore
+        collection = func(graph, *args)
+        return collection
+
+
+class NestedFrame(
+    _Frame, dd.DataFrame
+):  # can use dd.DataFrame instead of dx.DataFrame if the config is set true (default in >=2024.3.0)
+    """An extension for a Dask Dataframe that has Nested-Pandas functionality.
+
+    Examples
+    --------
+    >>> import nested_dask as nd
+    >>> base = nd.NestedFrame(base_data)
+    >>> layer = nd.NestedFrame(layer_data)
+    >>> base.add_nested(layer, "layer")
+    """
+
+    _partition_type = npd.NestedFrame  # Tracks the underlying data type
+
+    def __getitem__(self, key):
+        result = super().__getitem__(key)
+        return result
+
+    @classmethod
+    def from_nested_pandas(
+        cls,
+        data,
+        npartitions=None,
+        chunksize=None,
+        sort=True,
+    ) -> NestedFrame:
+        """Returns an Nested-Dask NestedFrame constructed from a Nested-Pandas
+        NestedFrame.
+
+        Parameters
+        ----------
+        data: `NestedFrame`
+            Nested-Pandas NestedFrame containing the underlying data
+        npartitions: `int`, optional
+            The number of partitions of the index to create. Note that depending on
+            the size and index of the dataframe, the output may have fewer
+            partitions than requested.
+        chunksize: `int`, optional
+            Size of the individual chunks of data in non-parallel objects that make up Dask frames.
```

Comment:
Is it size in counts or bytes? Should we make an issue to re-partition by byte size?

Reply:
docstring updated